### PR TITLE
Fix slave mode when no default layer is available

### PR DESCRIPTION
**Pull request by alexbrault**
This PR fixes some issues with slave mode related to the default layer (or lack thereof). The layer information is pulled from the correct layer wherever possible, or safely defaults to a lack of information.

### CODE REVIEW

**Comment by alexbrault** [comment]
The remaining failing tests appear to be unrelated to my changes

**Comment by sbrunner** [comment]
The Hash droppers and post process don't looks working ...
As you see I also do some changes on the base branch ...

**Comment by alexbrault** [comment]
There are still a few failing tests that I can't reproduce locally

**Comment by pvalsecc** [comment]
What do you mean by  "cannot reproduce locally"? You cannot run the tests locally or the tests are green on your machine?
I've tried on mine and they fail the same as on Travis.
And it's not a few tests, it more half of them that fail:
`......FF.........................F.......EE.F...EEFEEEEFFE.EEEFF.EEEEEE.EE`

**Comment by alexbrault** [comment]
I haven't been able to run the tests locally (mostly on account of my apache and postgres being configured for other purposes)

**Comment by pvalsecc** [comment]
I've added a few commits in this branch: https://github.com/camptocamp/tilecloud-chain/commits/pv_master_slave
To run the test, stop your local apache+postgresql and run that: `docker-compose -f test-docker-compose.yml up`

Then, you can run the tests with this command in another terminal: `.venv/bin/python setup.py nosetests --logging-filter=tilecloud,tilecloud_chain --attr '!'nopy`echo $TRAVIS_PYTHON_VERSION | awk -F . '{{print $1}}'``

I had to setup a venv with requirements.txt and dev-requirements.txt installed and I had to install optipng on my machine.

**Comment by alexbrault** [comment]
I've managed to get it down to 4 failing tests, but the ones that remain appear entirely unrelated to any of the work I've done for the slave mode

**Comment by pvalsecc** [comment]
I've had to fix a few tests because the sqs section more from layer to global. But I still have one failing test: `test_error_file in test_generate.py`
This test first fails because of a different accounting around ignored tiles and, if you change the expectations, it fails because the 2 expected tiles are not generated and 2 others are.
Can you have a look? I've tried and I don't understand what happens.

You can start from my branch which is based on yours: https://github.com/camptocamp/tilecloud-chain/tree/pv_master_slave

File: tilecloud_chain/__init__.py [hunk 1]
```diff
@@ -664,18 +666,33 @@ def log_tiles(tile):
...
+        store = MultiTileStore(splitters)
 
         if self.options.debug:
             def meta_get(tilestream):  # pragma: no cover
                 for metatile in tilestream:
                     substream = store.get((metatile,))
                     for tile in substream:
-                        tile.metatile = metatile
-                        tile.metadata = metatile.metadata
+                        if tile is not metatile:
```
Comment (sbrunner) [inline tilecloud_chain/__init__.py]: When is that true?

File: tilecloud_chain/__init__.py [hunk 2]
```diff
@@ -838,11 +856,22 @@ def counter(self, size=False):
...
+                self.imap(Process(self.config['process'][name], self.options))
+        else:
+            processes = {}
+            for lname, layer in self.layers.items():
+                name_ = name
+                if name_ is None:
+                    name_ = layer.get(key)
+                if name_ is not None:
+                    processes[lname] = Process(self.config['process'][name_], self.options)
+            if processes:
```
Comment (sbrunner) [inline tilecloud_chain/__init__.py]: Always True ...

File: tilecloud_chain/__init__.py [hunk 3]
```diff
@@ -838,11 +856,22 @@ def counter(self, size=False):
...
+        if self.layer:
+            if name is None:
+                name = self.layer.get(key)
+            if name is not None:
+                self.imap(Process(self.config['process'][name], self.options))
+        else:
+            processes = {}
+            for lname, layer in self.layers.items():
+                name_ = name
+                if name_ is None:
```
Comment (sbrunner) [inline tilecloud_chain/__init__.py]: When is it not None?

File: tilecloud_chain/__init__.py [hunk 4]
```diff
@@ -971,8 +1000,7 @@ def __init__(self):
         self.nb = 0
 
     def __call__(self, tile=None):
-        if tile and tile.data:
-            self.nb += 1
```
Comment (sbrunner) [inline tilecloud_chain/__init__.py]: This change looks strange ...

File: tilecloud_chain/__init__.py [hunk 5]
```diff
@@ -664,18 +666,33 @@ def log_tiles(tile):
...
-        store = MetaTileSplitterTileStore(
-            self.layer['mime_type'],
-            self.layer['grid_ref']['tile_size'],
-            self.layer['meta_buffer'])
+        class NullSplitter(TileStore):
+            @staticmethod
+            def get_one(tile):
+                return tile
+
+        splitters = {None: NullSplitter()}
```
Comment (sbrunner) [inline tilecloud_chain/__init__.py]: For what do we have this None key?

File: tilecloud_chain/generate.py [hunk 6]
```diff
@@ -107,7 +107,7 @@ def _gene(self, options, gene, dimensions=None):
         elif options.role == 'hash':
             try:
                 z, x, y = (int(v) for v in options.get_hash.split('/'))
-                if 'meta' in gene.layer:
+                if gene.layer.get('meta'):
```
Comment (sbrunner) [inline tilecloud_chain/generate.py]: Why doing this change?

File: tilecloud_chain/generate.py [hunk 7]
```diff
@@ -141,24 +141,37 @@ def wrong_content_type_to_error(tile):
...
-                            tile.data
+                            tile.error
                         )
                 return tile
             gene.imap(wrong_content_type_to_error)
+            gene.add_error_filters()
 
             if options.role == 'hash':
-                if 'meta' in gene.layer:
+                if gene.layer.get('meta', False):
```
Comment (sbrunner) [inline tilecloud_chain/generate.py]: Why doing this change?

**Comment by pvalsecc** [comment]
Thanks for the fix. python3 tests are is still failing, I'm on it.

File: tilecloud_chain/generate.py [hunk 8]
```diff
@@ -141,24 +141,37 @@ def wrong_content_type_to_error(tile):
...
+                            count=self.count_metatiles_dropped,
+                        )
+                if droppers:
+                    gene.imap(MultiHashDropper(droppers))
             elif not options.near:
                 # Discard tiles with certain content
-                if 'empty_tile_detection' in gene.layer:
-                    empty_tile = gene.layer['empty_tile_detection']
+                if gene.layer and 'empty_metatile_detection' in gene.layer:
+                    empty_tile = gene.layer['empty_metatile_detection']
```
Comment (sbrunner) [inline tilecloud_chain/generate.py]: Why changing tile to metatile?

File: tilecloud_chain/generate.py [hunk 9]
```diff
@@ -176,21 +189,34 @@ def add_elapsed_togenerate(metatile):
             gene.add_metatile_splitter()
             gene.imap(Logger(logger, logging.INFO, '%(tilecoord)s'))
 
-            if gene.layer['type'] != 'mapnik' or gene.layer['output_format'] != 'grid':
+            if self.count_tiles is None:
```
Comment (sbrunner) [inline tilecloud_chain/generate.py]: Why doing this change?

File: tilecloud_chain/generate.py [hunk 10]
```diff
@@ -176,21 +189,34 @@ def add_elapsed_togenerate(metatile):
...
                 gene.imap(HashLogger('empty_tile_detection'))
+            elif options.role == 'slave':
+                droppers = {}
+                for lname, layer in gene.layers.items():
+                    if 'empty_tile_detection' in layer:
+                        empty_tile = layer['empty_tile_detection']
+                        droppers[lname] = HashDropper(
+                            empty_tile['size'], empty_tile['hash'], store=self.cache_tilestore,
+                            queue_store=self.sqs_tilestore,
+                            count=self.count_metatiles_dropped,
```
Comment (sbrunner) [inline tilecloud_chain/generate.py]: metatile?

File: tilecloud_chain/tests/test_cost.py [hunk 11]
```diff
@@ -174,30 +174,35 @@ def test_cost_line(self):
                     'time': '0:00:00',
                     's3': '0.00',
                 },
+                'SQS usage: 0.00 [$]',
```
Comment (sbrunner) [inline tilecloud_chain/tests/test_cost.py]: This looks like hidding an other error ...

File: tilecloud_chain/tests/test_error.py [hunk 12]
```diff
@@ -158,7 +158,13 @@ def test_wrong_map(self, l):
...
         l.check((
             'tilecloud_chain', 'ERROR', """The config file 'tilegeneration/wrong_map.yaml' is invalid.
- - Value: test is not of a mapping type"""
+ - /: Cannot find required key 'caches'.
+ - /: Cannot find required key 'grids'.
+ - /layers/test/empty_tile_detection: Value 'test' is not a dict. Value path: '/layers/test/empty_tile_detection'
+ - /layers/test: Cannot find required key 'extension'.
+ - /layers/test: Cannot find required key 'grid'.
+ - /layers/test: Cannot find required key 'mime_type'.
+ - /layers/test: Cannot find required key 'wmts_style'."""
```
Comment (sbrunner) [inline tilecloud_chain/tests/test_error.py]: Why this change?

**Comment by pvalsecc** [comment]
OK, I've fixed python 3 in my branch: https://github.com/camptocamp/tilecloud-chain/tree/pv_master_slave

File: tilecloud_chain/__init__.py (hunk 5 above)
Comment (alexbrault) [inline tilecloud_chain/__init__.py]: This None key is used as the default splitter if the layer name is not found in the splitters dict

File: tilecloud_chain/__init__.py (hunk 1 above)
Comment (alexbrault) [inline tilecloud_chain/__init__.py]: Then the NullSplitter is used to "split" a tile from a layer with no metatile

File: tilecloud_chain/__init__.py [hunk 13]
```diff
@@ -666,11 +666,20 @@ def log_tiles(tile):
...
-            layer['meta_buffer'])
-            for lname, layer in self.layers.items()})
+        class NullSplitter(TileStore):
+            @staticmethod
+            def get_one(tile):
+                return tile
+
+        splitters = {None: NullSplitter()}
+        for lname, layer in self.layers.items():
+            if layer.get('meta'):
```
Comment (alexbrault) [inline tilecloud_chain/__init__.py]: if the layer has no meta, it will have no entry in the splitters dict and default to the NullSplitter

File: tilecloud_chain/__init__.py (hunk 3 above)
Comment (alexbrault) [inline tilecloud_chain/__init__.py]: In practice, never, but since the old version of this method had a name argument, I'm reluctant to remove it and potentially break custom chains

File: tilecloud_chain/__init__.py (hunk 2 above)
Comment (alexbrault) [inline tilecloud_chain/__init__.py]: If none of the layers have a process with a given key (e.g. if no layer has a post-processing step), processes will be falsy.

File: tilecloud_chain/generate.py (hunk 6 above)
Comment (alexbrault) [inline tilecloud_chain/generate.py]: Because `meta: off` in the yaml meant there was a 'meta' with a value of False. A presence check meant that layers with meta explicitly set to off were treated as if they had a metatile configuration

File: tilecloud_chain/generate.py (hunk 7 above)
Comment (alexbrault) [inline tilecloud_chain/generate.py]: See above

File: tilecloud_chain/__init__.py (hunk 4 above)
Comment (alexbrault) [inline tilecloud_chain/__init__.py]: This is reverting a change that broke the metatile counter (it counted before data was populated

File: tilecloud_chain/generate.py (hunk 8 above)
Comment (alexbrault) [inline tilecloud_chain/generate.py]: Because a typo in 49f5612dfbc69f12cf6d036ff27801a2483ba231 changed it to tile. This block was always meant to be the metatile dropper

File: tilecloud_chain/generate.py (hunk 9 above)
Comment (alexbrault) [inline tilecloud_chain/generate.py]: Because that's ultimately what the condition was checking­, except that checking if the counter is initialised is less brittle than checking if the conditions that would have led to the counter being initialised are false (especially in cases like slave mode where there is no layer in gene)

File: tilecloud_chain/generate.py (hunk 10 above)
Comment (alexbrault) [inline tilecloud_chain/generate.py]: Typo on my part. I'll fix it when pulling in @pvalsecc's changes for python3

File: tilecloud_chain/tests/test_error.py (hunk 12 above)
Comment (alexbrault) [inline tilecloud_chain/tests/test_error.py]: You'll have to ask @pvalsecc, this is his change

Comment (pvalsecc) [inline tilecloud_chain/tests/test_error.py]: We moved the SQS config at the global level. So now, every layer has SQS.

File: tilecloud_chain/tests/test_cost.py (hunk 11 above)
Comment (pvalsecc) [inline tilecloud_chain/tests/test_cost.py]: We moved the SQS config at the global level. So now, every layer has SQS.

File: tilecloud_chain/tests/test_error.py (hunk 12 above)
Comment (sbrunner) [inline tilecloud_chain/tests/test_error.py]: Having a look with @pvalsecc strange modification but looks good :-)

**Comment by eleu** [comment]
@sbrunner and @alexbrault could we merge this PR?

**Comment by sbrunner** [comment]
No there is too many duplicated and untested code ...

File: tilecloud_chain/tests/test_generate.py [hunk 14]
```diff
@@ -749,7 +749,7 @@ def test_mapnik_grid_drop(self, l):
                 tiles=((5, 7), (7, 4)),
                 regex=True,
                 expected="""The tile generation of layer 'mapnik_grid_drop' is finish
-Nb generated tiles: 12
+Nb generated tiles: 2
```
Comment (alexbrault) [inline tilecloud_chain/tests/test_generate.py]: Always instantiating all tile stores to remove duplication means either that having a Mapnik grid layer in the config file (even if it is not used in a generation) will create the counter so early that other layers will count their metatile with it, or that the counter is instantiated too late to count Mapnik tiles that were generated and dropped. The latter seemed a more acceptable tradeoff

File: tilecloud_chain/generate.py [hunk 15]
```diff
@@ -370,15 +349,14 @@ def _get_tilestore_and_message_for_layer(self, layer, gene):
...
                 headers=layer['headers'],
-            ), "Get tile from WMS")
+            )
         elif layer['type'] == 'mapnik':  # pragma: no cover
             from tilecloud.store.mapnik_ import MapnikTileStore
             from tilecloud_chain.mapnik_ import MapnikDropActionTileStore
 
             grid = gene.get_grid(layer['grid'])
             if layer['output_format'] == 'grid':
-                self.count_tiles = gene.counter()
```
Comment (sbrunner) [inline tilecloud_chain/generate.py]: Why this change?

**Comment by sbrunner** [comment]
Thanks @alexbrault looks much better :-)

File: tilecloud_chain/generate.py (hunk 15 above)
Comment (alexbrault) [inline tilecloud_chain/generate.py]: If it is left in, having a mapnik grid layer in the config file will break the counter for all other layers (it will count tiles before the splitter can do its job)

This does mean mapnik grid layers can't count their generated tiles correctly, but as I mentioned above, that seemed better than all other layers not counting theirs

Comment (sbrunner) [inline tilecloud_chain/generate.py]: Arrg :-/

Comment (sbrunner) [inline tilecloud_chain/generate.py]: The related change in the tests looks suspect, but it's not so important ...

File: tilecloud_chain/generate.py [hunk 16]
```diff
@@ -370,15 +349,14 @@ def _get_tilestore_and_message_for_layer(self, layer, gene):
...
-                return (MapnikDropActionTileStore(
+                return MapnikDropActionTileStore(
```
Comment (sbrunner) [inline tilecloud_chain/generate.py]: We should add a try catch block around this block to make the Travis tests pass...
